scipy/conftest.py

```diff
@@ -209,7 +210,7 @@ def num_parallel_threads():
     pytest.param(v, id=k, marks=pytest.mark.array_api_backends)
     for k, v in xp_available_backends.items()
 ])
-def xp(request):
+def xp(request, monkeypatch):
```

Comment:
This is going to interfere with [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel). Just noting for now. We don't really need to test jitting and free-threading safety at the same time, but test markers may start interfering with each other.

Reply:
We can disable all monkey-patching when we detect parallel running I suppose?

Reply:
Yes, that should work I hope.

I just tested, and can't make it fail locally. I think it'd need both free-threading and JAX, and that combo isn't yet available. And then testing like:
```
$ pixi r test-nogil -b jax.numpy -s _lib -- --parallel-threads=8
```

Reply:
You should definitely be able to make it fail with the GIL too. Probably your test data is small enough that the GIL is never released long enough for a context switch between two tests.

Reply:
Yeah, if I try really hard with heavy enough functions, it should be possible. Too much trouble though; this will be way easier when `cp313t` JAX arrives (soon hopefully).

Reply:
it should be as simple as

```python
def pytest.mark.parametrize("repeat", range(100))
def test1(xp, repeat):
    sleep(np.random.random() / 10)   # 0~100ms
    x = xp.asarray(...)
    y = jitted_function(x)
    xp_assert_equal(y, xp.asarray(...))
```

if [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel) doesn't parallelize different iterations of `parametrize`, you'll need two identical tests running in parallel.

xp_assert_equal should crash on xp!=jax.numpy saying that y has namespace jax.numpy.

Reply:
It works a bit differently, it runs a single pytest test, rewritten to run the same code in multiple threads within that one test (see https://github.com/Quansight-Labs/pytest-run-parallel/pull/24). It maximizes the chance of collisions - only in the absence of the GIL.